{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":5,"nbformat":4,"cells":[{"cell_type":"markdown","source":"![Variedades americanas](paper.jpg)","metadata":{},"id":"8febd1dd-74bd-44de-a5e4-1604a91d3c16"},{"cell_type":"markdown","source":"# Genómica poblacional del Mango en América","metadata":{},"id":"13ac633d-7057-4943-8e9c-7b7eb24a9477"},{"cell_type":"markdown","source":"### Prácticas de Bioinformática para el curso de Genómica de la Conservación 2022 de Ciencias Agrogenómicas\n### Asistente: Eddy Mendoza (eddyjmeg@gmail.com)","metadata":{},"id":"c454fc91-dd92-447c-8813-6328a4121cf3"},{"cell_type":"markdown","source":"### Tu nombre:","metadata":{},"id":"7415a918-3628-431e-a2d8-fb91e3638983"},{"cell_type":"markdown","source":"En esta segunda parte, exploraremos la variación genética y la estructura de las poblaciones usando los SNPs que llamamos anteriormente. \n\nComo habrás visto, la carpeta de salida tiene varios archivos. Nosotros usaremos <snps.vcf> y <.plink>\n\nEl software que usaremos en esta ocasión: \n\n    conda install -c bioconda plink\n    conda install -c bioconda admixture","metadata":{},"id":"d7f71274-5915-49d8-afc0-a4d2f3c31958"},{"cell_type":"markdown","source":"### Análisis de componentes principales","metadata":{},"id":"952887ab-0c88-423e-84ae-e5d6085d6202"},{"cell_type":"code","source":"# Primero necesitamos eliminar los SNPs que se encuentran ligados, estos nos pueden meter ruido.\n! sed 's/Vanilla_planifolia_SAGARPA_Sc//g' populations.snps.vcf > snps.vcf\n#Este comando los busca y los manda a otro archivo\n! plink --vcf snps.vcf --allow-extra-chr --indep-pairwise 50 10 0.1 --out linked\n\n# Ahora tomaremos el set original y le restaremos los SNPs ligados\n! echo \"___________________________________________\"\n! plink --vcf snps.vcf --allow-extra-chr --extract linked.prune.in --make-bed --mind --out pruned","metadata":{"trusted":true},"execution_count":1,"outputs":[{"name":"stdout","text":"PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to linked.log.\nOptions in effect:\n  --allow-extra-chr\n  --indep-pairwise 50 10 0.1\n  --out linked\n  --vcf snps.vcf\n\n772642 MB RAM detected; reserving 386321 MB for main workspace.\n--vcf: linked-temporary.bed + linked-temporary.bim + linked-temporary.fam\nwritten.\n59 variants loaded from .bim file.\n37 people (0 males, 0 females, 37 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to linked.nosex .\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 37 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.893724.\n59 variants and 37 people pass filters and QC.\nNote: No phenotypes present.\nPruned 0 variants from chromosome 27, leaving 1.\nPruned 0 variants from chromosome 28, leaving 1.\nPruned 1 variant from chromosome 29, leaving 1.\nPruned 0 variants from chromosome 30, leaving 1.\nPruned 0 variants from chromosome 31, leaving 1.\nPruned 0 variants from chromosome 32, leaving 1.\nPruned 0 variants from chromosome 33, leaving 1.\nPruned 2 variants from chromosome 34, leaving 1.\nPruned 0 variants from chromosome 35, leaving 1.\nPruned 1 variant from chromosome 36, leaving 1.\nPruned 0 variants from chromosome 37, leaving 1.\nPruned 0 variants from chromosome 38, leaving 1.\nPruned 0 variants from chromosome 39, leaving 1.\nPruned 1 variant from chromosome 40, leaving 1.\nPruned 0 variants from chromosome 41, leaving 1.\nPruned 1 variant from chromosome 42, leaving 1.\nPruned 1 variant from chromosome 43, leaving 1.\nPruned 0 variants from chromosome 44, leaving 1.\nPruned 1 variant from chromosome 45, leaving 1.\nPruned 0 variants from chromosome 46, leaving 1.\nPruned 1 variant from chromosome 47, leaving 1.\nPruned 0 variants from chromosome 48, leaving 1.\nPruned 0 variants from chromosome 49, leaving 1.\nPruned 0 variants from chromosome 50, leaving 1.\nPruned 0 variants from chromosome 51, leaving 1.\nPruned 0 variants from chromosome 52, leaving 1.\nPruned 0 variants from chromosome 53, leaving 1.\nPruned 0 variants from chromosome 54, leaving 1.\nPruned 2 variants from chromosome 55, leaving 1.\nPruned 2 variants from chromosome 56, leaving 1.\nPruned 0 variants from chromosome 57, leaving 1.\nPruned 0 variants from chromosome 58, leaving 1.\nPruned 3 variants from chromosome 59, leaving 1.\nPruned 0 variants from chromosome 60, leaving 1.\nPruned 1 variant from chromosome 61, leaving 1.\nPruned 0 variants from chromosome 62, leaving 1.\nPruned 0 variants from chromosome 63, leaving 1.\nPruned 0 variants from chromosome 64, leaving 1.\nPruned 0 variants from chromosome 65, leaving 1.\nPruned 0 variants from chromosome 66, leaving 1.\nPruned 0 variants from chromosome 67, leaving 1.\nPruned 0 variants from chromosome 68, leaving 1.\nPruning complete.  17 of 59 variants removed.\nMarker lists written to linked.prune.in and linked.prune.out .\n___________________________________________\nPLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to pruned.log.\nOptions in effect:\n  --allow-extra-chr\n  --extract linked.prune.in\n  --make-bed\n  --mind\n  --out pruned\n  --vcf snps.vcf\n\n772642 MB RAM detected; reserving 386321 MB for main workspace.\n--vcf: pruned-temporary.bed + pruned-temporary.bim + pruned-temporary.fam\nwritten.\n59 variants loaded from .bim file.\n37 people (0 males, 0 females, 37 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to pruned.nosex .\n--extract: 42 variants remaining.\n6 people removed due to missing genotype data (--mind).\nIDs written to pruned.irem .\nUsing 1 thread (no multithreaded calculations invoked).\nBefore main variant filters, 31 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate in remaining samples is 0.963134.\n42 variants and 31 people pass filters and QC.\nNote: No phenotypes present.\n--make-bed to pruned.bed + pruned.bim + pruned.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n","output_type":"stream"}],"id":"4976cc56-e3c0-4b7d-977f-586da4ff19d8"},{"cell_type":"code","source":"# Ahora le pediremos a PLINK que haga por nosotros un PCA\n! plink --bfile pruned --allow-extra-chr --pca --out pca","metadata":{"trusted":true},"execution_count":2,"outputs":[{"name":"stdout","text":"PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\nLogging to pca.log.\nOptions in effect:\n  --allow-extra-chr\n  --bfile pruned\n  --out pca\n  --pca\n\n772642 MB RAM detected; reserving 386321 MB for main workspace.\n42 variants loaded from .bim file.\n31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.\nAmbiguous sex IDs written to pca.nosex .\nUsing up to 71 threads (change this with --threads).\nBefore main variant filters, 31 founders and 0 nonfounders present.\nCalculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\nTotal genotyping rate is 0.963134.\n42 variants and 31 people pass filters and QC.\nNote: No phenotypes present.\nRelationship matrix calculation complete.\n--pca: Results saved to pca.eigenval and pca.eigenvec .\n","output_type":"stream"}],"id":"4cbe802c-1336-411a-b811-460d6c8fa034"},{"cell_type":"markdown","source":"### ADMIXTURE","metadata":{},"id":"85bf60cb-1889-4f40-993b-754bf726c521"},{"cell_type":"code","source":"! bash admixture.sh","metadata":{"trusted":true},"execution_count":19,"outputs":[],"id":"0d180937-8769-4076-bcd5-e44979d00ee5"},{"cell_type":"markdown","source":"### Ahora crea una sesión de R para graficar los resultados. Vamos a aprovechar a ponerle el nombre de las poblacioneas a cada individuo.  ","metadata":{},"id":"ba13783f-779d-42c2-96e2-efa4b4cfaa58"},{"cell_type":"code","source":"pip install pandas","metadata":{"trusted":true},"execution_count":21,"outputs":[{"name":"stdout","text":"Collecting pandas\n  Downloading pandas-1.3.5-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.3 MB)\n     |████████████████████████████████| 11.3 MB 8.0 MB/s            \n\u001b[?25hRequirement already satisfied: pytz>=2017.3 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from pandas) (2021.3)\nCollecting numpy>=1.17.3\n  Downloading numpy-1.21.5-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)\n     |████████████████████████████████| 15.7 MB 46.5 MB/s            \n\u001b[?25hRequirement already satisfied: python-dateutil>=2.7.3 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from pandas) (2.8.2)\nRequirement already satisfied: six>=1.5 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from python-dateutil>=2.7.3->pandas) (1.16.0)\nInstalling collected packages: numpy, pandas\nSuccessfully installed numpy-1.21.5 pandas-1.3.5\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}],"id":"db952b17-dc04-41fe-a7a1-7eaa951daeda"},{"cell_type":"code","source":"import pandas as pd\nimport os\nmetadata = pd.read_csv(\"metadata_sra.tsv\", sep='\\t', header=None)\npca = pd.read_csv(\"pca.eigenvec\", sep=' ', header=None)\nclusters = pd.merge(metadata, pca, left_on=3, right_on=0)","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"29836fbb-cfd2-47bf-b9b0-3f1323aa329c"}]}